lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -4046,6 +3978,134 @@ isExpressionEmittedInlineIntoProceduralDeclaration(Operation *op,
+  AssignTy singleAssign;
+  if (llvm::all_of(op->getUsers(), [&](Operation *user) {
+        if (hasSVAttributes(user))
+          return false;
+
+        if (auto assign = dyn_cast<AssignTy>(user)) {
+          if (singleAssign)
+            return false;
+          singleAssign = assign;
+          return true;
+        }
+
+        return isa<ReadInOutOp>(user);
+      }))
+    return singleAssign;
+  return {};
+}
+
+// Return true if `op1` dominates users of `op2`.
+static bool checkDominanceOfUsers(Operation *op1, Operation *op2) {
+  return llvm::all_of(op2->getUsers(), [&](Operation *user) {
+    if (op1->getBlock() != user->getBlock())
+      return false;
+
+    if (op1 == user)
+      return true;
+
+    return op1->isBeforeInBlock(user);
+  });
+}
+
+LogicalResult StmtEmitter::emitDeclaration(Operation *op) {
+  // If spillWiresAtPrepare option is disabled, use the previous emission
+  // method.
+  // TODO: Once ExportVerilog simplification finished, remove this condition.
+  if (!state.options.spillWiresAtPrepare)
+    return emitNoop();
+
+  emitSVAttributes(op);
+  auto value = op->getResult(0);
+  SmallPtrSet<Operation *, 8> opsForLocation;
+  opsForLocation.insert(op);
+
+  // Emit the leading word, like 'wire', 'reg' or 'logic'.
+  auto type = value.getType();
+  auto word = getVerilogDeclWord(op, state.options);
+  if (!isZeroBitType(type)) {
+    indent() << word;
+    auto extraIndent = word.empty() ? 0 : 1;
+    os.indent(maxDeclNameWidth - word.size() + extraIndent);
+  } else {
+    indent() << "// Zero width: " << word << ' ';
+  }
+
+  SmallString<8> typeString;
+  // Convert the port's type to a string and measure it.
+  {
+    llvm::raw_svector_ostream stringStream(typeString);
+    emitter.printPackedType(stripUnpackedTypes(type), stringStream,
+                            op->getLoc());
+  }
+  // Emit the type.
+  os << typeString;
+  if (typeString.size() < maxTypeWidth)
+    os.indent(maxTypeWidth - typeString.size());
+
+  // Emit the name.
+  os << names.getName(value);
+
+  // Print out any array subscripts or other post-name stuff.
+  emitter.printUnpackedTypePostfix(type, os);
+
+  if (auto localparam = dyn_cast<LocalParamOp>(op)) {
+    os << " = ";
+    emitter.printParamValue(localparam.getValue(), os, [&]() {
+      return op->emitOpError("invalid localparam value");
+    });
+  }
+
+  // Try inlining an assignment into declarations.
+  if (isa<WireOp, LogicOp>(op) &&
+      !op->getParentOp()->hasTrait<ProceduralRegion>()) {
+    // Get a single assignments if any.
+    if (auto singleAssign = getSingleAssignAndCheckUsers<AssignOp>(op)) {
+      auto source = singleAssign.getSrc().getDefiningOp();
+      // Check that the source value is OK to inline in the current emission
+      // point. A port or constant is fine, otherwise check that the assign is
+      // next to the operation.
+      if (!source || isa<ConstantOp>(source) ||
+          op->getNextNode() == singleAssign) {
```

Comment:
Just for my understanding, this is another practical simplification that is useful already, but a more precise check could be implemented?

Reply:
Yes, we can perform analysis similar to `isExpressionEmittedInlineIntoProceduralDeclaration` if we want. `op->getNextNode() == singleAssign` is a simple way to guarantee that the source value of assign can be inlined at the assign op.